src/v/storage/disk_log_impl.cc

```diff
@@ -1970,26 +1970,112 @@ disk_log_impl::disk_usage_target(gc_config cfg, usage usage) {
+        target.min_capacity_wanted = std::min(
+          want_time.value(), want_size.value());
+    } else if (want_time.has_value()) {
+        target.min_capacity_wanted = want_time.value();
+    } else {
+        target.min_capacity_wanted = want_size.value();
     }
 
     co_return target;
 }
 
+ss::future<std::optional<size_t>>
+disk_log_impl::disk_usage_target_time_retention(gc_config cfg) {
+    /*
+     * take the opportunity to maybe fixup janky weird timestamps. also done in
+     * normal garbage collection path and should be idempotent.
+     */
+    if (auto ignore_timestamps_in_future
+        = config::shard_local_cfg().storage_ignore_timestamps_in_future_sec();
+        ignore_timestamps_in_future.has_value()) {
+        co_await retention_adjust_timestamps(
+          ignore_timestamps_in_future.value());
+    }
+
+    /*
+     * we are going to use whole segments for the data we'll use to extrapolate
+     * the time-based retention capacity needs to avoid complexity of index
+     * lookups (for now); if there isn't an entire segment worth of data, we
+     * might not have a good sample size either. first we'll find the oldest
+     * segment whose starting offset is greater than the eviction time. this and
+     * subsequent segments will be fully contained within the time range.
+     */
+    auto it = std::find_if(
+      std::cbegin(_segs),
+      std::cend(_segs),
+      [time = cfg.eviction_time](const ss::lw_shared_ptr<segment>& s) {
+          return s->index().base_timestamp() >= time;
+      });
+
+    // collect segments for reducing
+    fragmented_vector<segment_set::type> segments;
+    for (; it != std::cend(_segs); ++it) {
+        segments.push_back(*it);
+    }
+
+    /*
+     * not enough data. caller will substitute in a reasonable value.
+     */
+    if (segments.size() < 2) {
+        co_return std::nullopt;
+    }
+
+    auto start_timestamp = segments.front()->index().base_timestamp();
+    auto end_timestamp = segments.back()->index().max_timestamp();
+    auto duration = end_timestamp - start_timestamp;
+    auto missing = start_timestamp - cfg.eviction_time;
+
+    /*
+     * timestamps are weird or there isn't a lot of data. be careful with "not a
+     * lot of data" when considering time because throughput may be large. so we
+     * go with something like 10 seconds just as a sanity check.
+     */
+    if (
+      missing <= model::timestamp(0) || duration <= model::timestamp(10'000)) {
+        co_return std::nullopt;
+    }
+
+    // roll up the amount of disk space taken by these segments
+    auto usage = co_await ss::map_reduce(
+      segments,
+      [](const segment_set::type& seg) { return seg->persistent_size(); },
+      storage::usage{},
+      [](storage::usage acc, storage::usage u) { return acc + u; });
+
+    // extrapolate out for the missing period of time in the retention period
+    auto missing_bytes = (usage.total() * missing.value()) / duration.value();
```

Comment:
I think this is saying "I expect to see same throughput that ingested the latest segments for the entire remainder of the retention period". This seems like a reasonable, but it makes me wonder about bursty workloads, where a large number of records are written in the span of a few dozens of seconds, and then stopped. In such a case, we could end up with a small `duration` and large `missing` and end up with a very large value for `missing_bytes`.

Maybe it makes sense to compute the expected throughput with `now() - start_timestamp` instead of `end_timestamp - start_timestamp`, though given these are user-input timestamps, it's probably best to avoid using `now()`

I suppose regardless it might be better to overestimate the wanted bytes anyway for the goal of avoiding out of space issues

Reply:
I attempted to capture this issue with a heuristic like not making any estimate unless we have say 10 seconds worth of data. But this is going to just be a whack-a-mole game I'm afraid. Perhaps we would do well to do add a cap here, too, such as double or triple the amount of data currently on disk. This may work well since the goal isn't necessarily to observe the system once and know how much space we need. Rather, this will be monitoring periodically. So capping will let the estimate continue to portray growth requirement, while also eliminating huge bursts?